File: Arboles R-Arios/codigo1.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Nodo {
    int valor;
    struct Nodo* xor_link;
    struct Nodo* siguienteHijo;  // Usado para hijos adicionales en la lista ligada.
} Nodo;

Nodo* XOR(Nodo* a, Nodo* b) {
    return (Nodo*)((size_t)a ^ (size_t)b);
}

Nodo* crearNodo(int valor) {
    Nodo* nuevo = (Nodo*)malloc(sizeof(Nodo));
    nuevo->valor = valor;
    nuevo->xor_link = NULL;
    nuevo->siguienteHijo = NULL;
    return nuevo;
}

void agregarHijo(Nodo* padre, Nodo* hijo) {
    if (!padre->xor_link) {
        padre->xor_link = hijo;
    } else {
        Nodo* primerHijo = padre->xor_link;
        while (primerHijo->siguienteHijo) {
            primerHijo = primerHijo->siguienteHijo;
        }
        primerHijo->siguienteHijo = hijo;
        hijo->xor_link = XOR(hijo->xor_link, padre);  // Apunta al padre.
    }
}

void imprimirHijos(Nodo* padre) {
    Nodo* hijoActual = padre->xor_link;
    while (hijoActual) {
        printf("Hijo: %d\n", hijoActual->valor);
        hijoActual = hijoActual->siguienteHijo;
    }
}

int main() {
    Nodo* raiz = crearNodo(10);
    Nodo* hijo1 = crearNodo(5);
    Nodo* hijo2 = crearNodo(6);
    Nodo* hijo3 = crearNodo(7);

    agregarHijo(raiz, hijo1);
    agregarHijo(raiz, hijo2);
    agregarHijo(raiz, hijo3);

    printf("Raíz: %d\n", raiz->valor);
    imprimirHijos(raiz);

    return 0;
}

File: Arboles Binarios/1-Version JavaScript/codigo1.js
class Nodo {
    constructor(valor) {
        this.valor = valor;
        this.link = null;  // XOR del hijo y hermano
    }
}

function XOR(a, b) {
    return a ^ b;
}

function crearNodo(valor) {
    return new Nodo(valor);
}

function agregarHijo(padre, hijo) {
    if (!padre.link) {
        padre.link = hijo;
    } else {
        let hermano = padre.link;
        padre.link = XOR(hermano, hijo);
        hijo.link = XOR(hijo.link, hermano);
    }
}

function contarNodos(nodo, prev = 0) {
    if (!nodo) return 0;

    let next = XOR(prev, nodo.link);

    return 1 + contarNodos(nodo.link, nodo) + contarNodos(next, nodo);
}

function altura(nodo, prev = 0) {
    if (!nodo) return 0;

    let next = XOR(prev, nodo.link);
    let h1 = 1 + altura(nodo.link, nodo);
    let h2 = 1 + altura(next, nodo);
    
    return Math.max(h1, h2);
}

// Prueba
let raiz = crearNodo(10);
let hijo1 = crearNodo(5);
let hijo2 = crearNodo(6);
let hijo3 = crearNodo(7);

agregarHijo(raiz, hijo1);
agregarHijo(hijo1, hijo2);
agregarHijo(hijo2, hijo3);

console.log(`Raíz: ${raiz.valor}`);
console.log(`Nodos totales: ${contarNodos(raiz)}`);
console.log(`Altura del árbol: ${altura(raiz)}`);
